src/tokens2.jl

```diff
@@ -0,0 +1,179 @@
+
+@inline startof(m::SAContainer) = IntSemiToken(beginloc(m.bt))
+
+## Function endof returns the semitoken that points
+## to the last item in the sorted order,
+## or the before-start marker if the tree is empty.
+
+@inline endof(m::SAContainer) = IntSemiToken(endloc(m.bt))
+
+## Function pastendsemitoken returns the token past the end of the data.
+
+@inline pastendsemitoken(::SAContainer) = IntSemiToken(2)
+
+## Function beforestarttoken returns the token before the start of the data.
+
+@inline beforestartsemitoken(::SAContainer) = IntSemiToken(1)
+
+## delete! deletes an item given a token.
+
+@inline function delete!(ii::Token)
+    has_data(ii)
+    delete!(ii[1].bt, ii[2].address)
+end
+
+## Function advances takes a token and returns the
+## next token in the sorted order. 
+
+@inline function advance(ii::Token)
+    not_pastend(ii)
+    IntSemiToken(nextloc0(ii[1].bt, ii[2].address))
+end
+
+
+## Function regresss takes a token and returns the
+## previous token in the sorted order. 
+
+@inline function regress(ii::Token)
+    not_beforestart(ii)
+    IntSemiToken(prevloc0(ii[1].bt, ii[2].address))
+end
+
+
+## status of a token is 0 if the token is invalid, 1 if it points to 
+## ordinary data, 2 if it points to the before-start location and 3 if
+## it points to the past-end location.
+
+
+@inline status(ii::Token) =
+       !(ii[2].address in ii[1].bt.useddatacells)? 0 :
+         ii[2].address == 1?                       2 : 
+         ii[2].address == 2?                       3 : 1
+
+
+@inline compare(m::SAContainer, s::IntSemiToken, t::IntSemiToken) = 
+      compareInd(m.bt, s.address, t.address)
+
+
+@inline function deref(ii::SDMToken)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k, ii[1].bt.data[ii[2].address].d
+end
+
+@inline function deref(ii::SetToken)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k
+end
+
+@inline function deref_key(ii::SDMToken)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].k
+end
+
+@inline function deref_value(ii::SDMToken)
+    has_data(ii)
+    return ii[1].bt.data[ii[2].address].d
+end
+
+
+
+## Functions setindex! and getindex for semitokens.
+## Note that we can't use SDMContainer here; we have
+## to spell it out otherwise there is an ambiguity.
+
+@inline function getindex(m::SortedDict,
+                          i::IntSemiToken)    
+    has_data((m,i))
+    return m.bt.data[i.address].d
+end
+
+@inline function getindex(m::SortedMultiDict,
+                          i::IntSemiToken)    
+    has_data((m,i))
+    return m.bt.data[i.address].d
+end
+
+@inline function setindex!(m::SortedDict,
+                           d_, 
+                           i::IntSemiToken)
+    has_data((m,i))
+    m.bt.data[i.address] = KDRec{keytype(m),datatype(m)}(m.bt.data[i.address].parent,
+                                                         m.bt.data[i.address].k, 
+                                                         convert(datatype(m),d_))
```

Comment:
Just curious, is there a reason you chose to use `keytype` and `datatype` instead of writing this as (the more standard)

```julia
@inline function setindex!{K,V}(m::SortedDict{K,V},
                                d_,
                                i::IntSemiToken)
     has_data((m,i))
     m.bt.data[i.address] = KDRec{K,V}(m.bt.data[i.address].parent,
                                       m.bt.data[i.address].k, 
                                       convert(V,d_))
```

Either way, `setindex!` would be specialized for the different `K` and `V` types, (and there's no need to capture the `Ordering` if you don't use it).

Would you mind changing this (and the other occurrences indicated below) to this style?